File: vol1/c-examples/RBFNetwork.c
#include "aifh-vol1.h"


RBF_NETWORK* RBFNetworkCreate(RBF_FUNCTION rbf, int input_count, int rbf_count, int output_count) {
	RBF_NETWORK *result;
	int i;

	/* calculate input and output weight counts
	   add 1 to output to account for an extra bias node */
	const int inputWeightCount = input_count * rbf_count;
	const int outputWeightCount = (rbf_count + 1) * output_count;
	const int rbfParams = (input_count + 1) * rbf_count;

	/* allocate structure */
	result = (RBF_NETWORK*)calloc(1,sizeof(RBF_NETWORK));
	result->input_count = input_count;
	result->rbf_count = rbf_count;
	result->rbf = rbf;
	result->output_count = output_count;
	result->rbf_index = (unsigned int*)calloc(rbf_count,sizeof(int));
	result->rbf_output = (double*)calloc(rbf_count+1,sizeof(double));
	result->weighted_input = (double*)calloc(input_count,sizeof(double));

	result->ltm_size = inputWeightCount + outputWeightCount + rbfParams;
	result->long_term_memory = (double*)calloc(result->ltm_size,sizeof(double));

	result->index_input_weights = 0;
	result->index_output_weights = inputWeightCount + rbfParams;

	for (i = 0; i < rbf_count; i++) {
		result->rbf_index[i] = inputWeightCount + ((input_count + 1) * i);
	}

	return result;
}

void RBFNetworkDelete(RBF_NETWORK *network) {
	free(network->rbf_index);
	free(network->rbf_output);
	free(network->weighted_input);
	free(network);
}

void RBFNetworkReset(RBF_NETWORK *network) {
	unsigned int i;
	RANDOM_GENERATE *prng;

	prng = RandCreate(TYPE_RANDOM_MT,(long)time(NULL));
	for(i=0;i<network->ltm_size;i++) {
		network->long_term_memory[i] = RandNextDoubleRange(prng,-1,1);
	}

	RandDelete(prng);
}


void RBFNetworkComputeRegression(RBF_NETWORK *network, double *input, double *output) {
	unsigned int inputIndex, outputIndex, rbfIndex, memoryIndex;
	double sum;

	/* first, compute the output values of each of the RBFs
	   Add in one additional RBF output for bias (always set to one). */

	network->rbf_output[network->rbf_count] = 1; // bias

	for (rbfIndex = 0; rbfIndex < network->rbf_count; rbfIndex++) {
		/* weight the input */
		for (inputIndex = 0; inputIndex < network->input_count; inputIndex++) {
			memoryIndex = network->index_input_weights + (rbfIndex * network->input_count) + inputIndex;
			network->weighted_input[inputIndex] = input[inputIndex] * network->long_term_memory[memoryIndex];
		}

		/* calculate the rbf */
		network->rbf_output[rbfIndex] = network->rbf(network->weighted_input,0,network->input_count,network->long_term_memory,network->rbf_index[rbfIndex]);
	}

	/* second, calculate the output, which is the result of the weighted result of the RBF's. */

	for (outputIndex = 0; outputIndex < network->output_count; outputIndex++) {
		sum = 0;
		for (rbfIndex = 0; rbfIndex <= network->rbf_count; rbfIndex++) {
			/* add 1 to rbf length for bias */
			memoryIndex = network->index_output_weights + (outputIndex * (network->rbf_count + 1)) + rbfIndex;
			sum += network->rbf_output[rbfIndex] * network->long_term_memory[memoryIndex];
		}
		output[outputIndex] = sum;
	}
}

File: vol1/scala-examples/src/main/scala/com/heatonresearch/aifh/learning/RBFNetwork.scala
/*
* Artificial Intelligence for Humans
* Volume 1: Fundamental Algorithms
* Scala Version
* http://www.aifh.org
* http://www.jeffheaton.com
*
* Code repository:
* https://github.com/jeffheaton/aifh

* Copyright 2013 by Jeff Heaton
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
* For more information on Heaton Research copyrights, licenses
* and trademarks visit:
* http://www.heatonresearch.com/copyright
*/
package com.heatonresearch.aifh.learning

import com.heatonresearch.aifh.general.fns.FnRBF
import com.heatonresearch.aifh.general.fns.GaussianFunction
import com.heatonresearch.aifh.randomize.GenerateRandom
import com.heatonresearch.aifh.general.data.RichData
import RichData._
import scala.collection.mutable.ArrayBuffer

/**
 * A RBF network is an advanced machine learning algorithm that uses a series of RBF functions to perform
 * regression.  It can also perform classification by means of one-of-n encoding.
 * <p/>
 * The long term memory of a RBF network is made up of the widths and centers of the RBF functions, as well as
 * input and output weighting.
 * <p/>
 * http://en.wikipedia.org/wiki/RBF_network
 *
 * Construct the RBF network.
 *
 * @param inputCount  The input count.
 * @param rbfCount       The number of RBF functions.
 * @param outputCount The output count.
 */
class RBFNetwork(val inputCount: Int, rbfCount: Int, val outputCount: Int) extends RegressionAlgorithm with ClassificationAlgorithm {

  /**
   * An index to the input weights in the long term memory.
   */
  private val indexInputWeights = 0

  val inputWeightCount = inputCount * rbfCount
  val outputWeightCount = (rbfCount + 1) * outputCount
  val rbfParams = (inputCount + 1) * rbfCount
  /**
   * An index to the output weights in the long term memory.
   */
  private val indexOutputWeights: Int = inputWeightCount + rbfParams

  /**
   * The weights & RBF parameters.  See constructor for layout.
   */
  override val longTermMemory = ArrayBuffer.fill(inputWeightCount + outputWeightCount + rbfParams)(0.0)

  /**
   * The RBF functions.
   */
  private val rbf: Vector[FnRBF] = {
    val arr = Array.ofDim[FnRBF](rbfCount)
    for(i <- 0 until rbfCount) {
      val rbfIndex: Int = inputWeightCount + ((inputCount + 1) * i)
      arr(i) = new GaussianFunction(inputCount, longTermMemory, rbfIndex)
    }
    arr.toVector
  }

  override def computeRegression(input: Vector[Double]): Vector[Double] = {
    val rbfOutput = ArrayBuffer.fill(rbf.length + 1)(0.0)
    rbfOutput(rbfOutput.length - 1) = 1.0
    for(rbfIndex <- 0 until rbf.length) {
      val weightedInput = for(inputIndex <- 0 until input.length) yield {
        val memoryIndex = indexInputWeights + (rbfIndex * inputCount) + inputIndex
        input(inputIndex) * longTermMemory(memoryIndex)
      }
      rbfOutput(rbfIndex) = rbf(rbfIndex).evaluate(weightedInput.toVector)

    }

    val result = ArrayBuffer.fill(outputCount)(0.0)
    for(outputIndex <- 0 until result.length) {
      var sum = 0.0
      for(rbfIndex <- 0 until rbfOutput.length) {
        val memoryIndex = indexOutputWeights + (outputIndex * (rbf.length + 1)) + rbfIndex
        sum += rbfOutput(rbfIndex) * longTermMemory(memoryIndex)
      }
      result(outputIndex) = sum
    }

    result.toVector
  }

  /**
   * Randomize the long term memory, with the specified random number generator.
   *
   * @param rnd A random number generator.
   */
  def reset(rnd: GenerateRandom) {
    for(i <- 0 until longTermMemory.length)
    longTermMemory(i) = rnd.nextDouble(-1, 1)
  }

  override def computeClassification(input: Vector[Double]): Int = {
    val output = computeRegression(input)
    output.maxIndex
  }

  override def toString: String = {
    s"[RBFNetwork:inputCount=$inputCount,outputCount=$outputCount,RBFs=${rbf.mkString("[",",","]")}]"
  }
}
